bin/propolis-server/src/lib/migrate/source.rs

```diff
@@ -137,17 +147,47 @@ impl<T: AsyncRead + AsyncWrite + Unpin + Send> SourceProtocol<T> {
         self.read_ok().await
     }
 
-    async fn ram_push(&mut self) -> Result<(), MigrateError> {
-        self.update_state(MigrationState::RamPush).await;
+    async fn ram_push(
+        &mut self,
+        phase: &MigratePhase,
+    ) -> Result<(), MigrateError> {
+        // Only publish the "RAM push" phase before pausing (so that the
+        // migration doesn't go from the "pause" state back to the "RAM push"
+        // state).
+        if matches!(phase, MigratePhase::RamPushPrePause) {
+            self.update_state(MigrationState::RamPush).await;
+        }
+
         let vmm_ram_range = self.vmm_ram_bounds().await?;
         let req_ram_range = self.read_mem_query().await?;
         info!(
             self.log(),
-            "ram_push: got query for range {:?}, vm range {:?}",
+            "ram_push ({:?}): got query for range {:?}, vm range {:?}",
+            phase,
             req_ram_range,
             vmm_ram_range
         );
-        self.offer_ram(vmm_ram_range, req_ram_range).await?;
+
+        // TODO: Ideally, both the pre-pause and post-pause phases would offer
```

Comment:
File an issue for this?

Reply:
Filed #387 and updated the comment in 46f8f99.

Reply:
thanks!